test/e2e/scaffold/gateway.go

```diff
@@ -0,0 +1,85 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package scaffold
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	ginkgo "github.com/onsi/ginkgo/v2"
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	gatewayv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gatewayclientset "sigs.k8s.io/gateway-api/pkg/client/clientset/gateway/versioned"
+)
+
+var (
+	_udpRouteTemplate = `
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: UDPRoute
+metadata:
+  name: %s
+spec:
+  rules:
+  - backendRefs:
+    - name: %s
+      port: %d
+`
+)
+
+func (s *Scaffold) getGatewayClientset() (*gatewayclientset.Clientset, error) {
+	var err error
+	var config *rest.Config
+
+	if s.kubectlOptions.InClusterAuth {
+		config, err = rest.InClusterConfig()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		kubeConfigPath, err := s.kubectlOptions.GetConfigPath(s.t)
+		if err != nil {
+			return nil, err
+		}
+		config, err = k8s.LoadApiClientConfigE(kubeConfigPath, s.kubectlOptions.ContextName)
+		if err != nil {
+			config, err = rest.InClusterConfig()
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	clientset, err := gatewayclientset.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientset, nil
+
+}
+
+func (s *Scaffold) CreateUDPRoute(name string, backendName string, backendPort int32) *gatewayv1alpha2.UDPRoute {
+	udpRoute := fmt.Sprintf(_udpRouteTemplate, name, backendName, backendPort)
+	err := s.CreateResourceFromString(udpRoute)
+	assert.Nil(ginkgo.GinkgoT(), err, "create UDPRoute failed")
+	client, err := s.getGatewayClientset()
+	assert.Nil(ginkgo.GinkgoT(), err, "get GatewayClientset failed")
+	route, err := client.GatewayV1alpha2().UDPRoutes(s.namespace).Get(context.TODO(), name, metav1.GetOptions{})
+	assert.Nil(ginkgo.GinkgoT(), err, "get UDPRoute failed")
```

Comment:
Usually we don't check it, instead we check whether the corresponding resource is generated in APISIX

Reply:
I think the semantics of `CreateUDPRoute` is make sure create UDPRoute successfully. If we don't check the error here, then we need to return that error and change signature like: 
```
func (s *Scaffold) CreateUDPRoute(...) (*gatewayv1alpha2.UDPRoute, error)
```
And then the caller still need to check the error which `CreateUDPRoute` returned. Every test case which create UDPRoute need to check that error. And that will make our code more boilerplate.

So I suggest we keep this semantics. What do you think?

Reply:
sorry for delay! 

```go
        udpRoute := fmt.Sprintf(_udpRouteTemplate, name, backendName, backendPort)
	err := s.CreateResourceFromString(udpRoute)
	assert.Nil(ginkgo.GinkgoT(), err, "create UDPRoute failed")
```

In fact we have done error checking here.  So we don't need to do this repeatedly

Reply:
At the same time I don't think it is necessary to add `CreateUDPRoute` function, 
 `CreateResourceFromString` is enough, right?

Reply:
> At the same time I don't think it is necessary to add `CreateUDPRoute` function, `CreateResourceFromString` is enough, right?

The function `CreateUDPRoute` is mainly intended to be able to reuse code. Otherwise, other test-cases will need to copy-paste the YAML template when they need to create a `UDPRoute`.

Reply:
These templates are all replaced by passing variables, like:

```go
udpRoute := fmt.Sprintf(_udpRouteTemplate, name, backendName, backendPort)
```

I don't see a clear benefit from it

Reply:
But it won't be a key to blocking this PR merge.

The main thing at the moment is the need to resolve conflicts so we can move forward

Reply:
I got it. We can just use `udpRoute := fmt.Sprintf(_udpRouteTemplate, name, backendName, backendPort)` instead of `CreateUDPRoute`. And error checking should be the responsibility of the creator.